### ISSUE TITLE
STL compatible size(), dimensions() return size_t 

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#357]
A coding-style proposal: 
our new code will be STL compatible and `size()`, `dimensions()`, capacity, etc will return `size_t`
value; 

Current: 
we use UInt or certain type that is used in the container (eg `vector<char>()::size()` we'd cast to char. 

New: STL uses size_type (size_t, ulong) 

Motivation: in PR #331 I'm playing with changing back-end data-types for containers. We get incompatibilities with other parts in the code on each change, deciding on size_t would fix this.

**Comment by ctrl-z-9000-times** [comment]
This sounds great.  Its always nice to be compatible with the standard library.

**Comment by dkeeney** [comment]
I agree.  Anything related to size ... or indexes for that matter, should be size_t.   There is a nupic::Size type but it is not consistently used.

**Comment by breznak** [comment]
> There is a nupic::Size type but it is not consistently used. 

good point, should we use `nupic::Size` (i think now typedef'ed to size_t), or just drop it and deal with `size_t` only?

**Comment by dkeeney** [comment]
>  or just drop it and deal with size_t only?

Well, for what it is worth, Cereal uses an `unsigned long long int` for size rather than size_t.  But I think that is a spacial case and everything else uses size_t.

Let's just use size_t.

**Comment by breznak** [comment]
from https://github.com/htm-community/nupic.cpp/pull/372#issuecomment-480546242 , @dkeeney 

> To be consistent....we should make ALL size related things size_t. This includes the dimensions. 

yes, although I'd like to advocate against that. 
- making dimensions size_t turned out a pain, and I messed up with SDR, SdrTools
- more importantly, semantics and ranges:
  - dimensions dictate size of (number of) cells (CellIdx), so it makes sense dimensions be CellIdx
    - dimensions is an input arg, CellIdx does not allow the user to set a number out of range
  - size_t is (typically) used for return types (size_t getNumSynapses()) with motivation to allow toggling types in Connections
